File: OximyWindows/src/OximyWindows/Services/UpdateCheckService.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Serialization;
using OximyWindows.Core;

namespace OximyWindows.Services;

/// <summary>
/// Checks for app updates by fetching version.json from GitHub Releases.
/// Respects MDM-managed deployments by skipping the check entirely.
/// </summary>
public class UpdateCheckService
{
    private static UpdateCheckService? _instance;
    public static UpdateCheckService Instance => _instance ??= new UpdateCheckService();

    private const string VersionCheckUrl = "https://github.com/OximyHQ/sensor/releases/download/latest/version.json";

    private bool _hasChecked;

    public bool UpdateAvailable { get; private set; }
    public bool Unsupported { get; private set; }
    public string? LatestVersion { get; private set; }
    public string? DownloadUrl { get; private set; }

    public event EventHandler? UpdateStatusChanged;

    private UpdateCheckService() { }

    /// <summary>
    /// Check for updates once on app launch. Fails silently.
    /// </summary>
    public async Task CheckOnceAsync()
    {
        if (_hasChecked) return;
        _hasChecked = true;

        // Skip update checks for MDM-managed devices
        if (MDMConfigService.Instance.IsManagedDevice) return;

        try
        {
            await CheckAsync();
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"[UpdateCheck] Failed: {ex.Message}");
            // Fail silently — update check is best-effort
        }
    }

    private async Task CheckAsync()
    {
        using var client = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
        var response = await client.GetAsync(VersionCheckUrl);

        if (!response.IsSuccessStatusCode) return;

        var json = await response.Content.ReadAsStringAsync();
        var info = JsonSerializer.Deserialize<VersionInfo>(json);
        if (info == null) return;

        var currentVersion = Constants.Version;

        if (CompareVersions(currentVersion, info.MinSupported) < 0)
        {
            // Below minimum supported version
            LatestVersion = info.Latest;
            DownloadUrl = info.Download?.Windows;
            Unsupported = true;
            UpdateAvailable = true;
            UpdateStatusChanged?.Invoke(this, EventArgs.Empty);
        }
        else if (CompareVersions(currentVersion, info.Latest) < 0)
        {
            // Update available but not critical
            LatestVersion = info.Latest;
            DownloadUrl = info.Download?.Windows;
            UpdateAvailable = true;
            UpdateStatusChanged?.Invoke(this, EventArgs.Empty);
        }
    }

    /// <summary>
    /// Semantic version comparison. Returns negative if a &lt; b, 0 if equal, positive if a &gt; b.
    /// </summary>
    private static int CompareVersions(string a, string b)
    {
        var aParts = a.Split('.').Select(s => int.TryParse(s, out var n) ? n : 0).ToArray();
        var bParts = b.Split('.').Select(s => int.TryParse(s, out var n) ? n : 0).ToArray();
        var maxLen = Math.Max(aParts.Length, bParts.Length);

        for (int i = 0; i < maxLen; i++)
        {
            var aVal = i < aParts.Length ? aParts[i] : 0;
            var bVal = i < bParts.Length ? bParts[i] : 0;
            if (aVal != bVal) return aVal.CompareTo(bVal);
        }
        return 0;
    }

    private class VersionInfo
    {
        [JsonPropertyName("latest")]
        public string Latest { get; set; } = "";

        [JsonPropertyName("min_supported")]
        public string MinSupported { get; set; } = "";

        [JsonPropertyName("download")]
        public DownloadInfo? Download { get; set; }
    }

    private class DownloadInfo
    {
        [JsonPropertyName("macos")]
        public string? Macos { get; set; }

        [JsonPropertyName("windows")]
        public string? Windows { get; set; }
    }
}

File: OximyMac/Services/UpdateCheckService.swift
import Foundation

/// Checks for app updates by fetching version.json from GitHub Releases.
/// Respects MDM-managed deployments by skipping the check entirely.
@MainActor
final class UpdateCheckService: ObservableObject {
    static let shared = UpdateCheckService()

    @Published var updateAvailable = false
    @Published var unsupported = false
    @Published var latestVersion: String?
    @Published var downloadURL: URL?

    private static let versionCheckURL = URL(string: "https://github.com/OximyHQ/sensor/releases/download/latest/version.json")!

    private var hasChecked = false

    private init() {}

    /// Check for updates on app launch. Fails silently — never blocks the app.
    func checkOnce() {
        guard !hasChecked else { return }
        hasChecked = true

        // Skip update checks for MDM-managed devices
        if MDMConfigService.shared.isManagedDevice {
            return
        }

        Task {
            await check()
        }
    }

    private func check() async {
        var request = URLRequest(url: Self.versionCheckURL)
        request.timeoutInterval = 10
        request.cachePolicy = .reloadIgnoringLocalCacheData

        do {
            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                return
            }

            let info = try JSONDecoder().decode(VersionInfo.self, from: data)
            let currentVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"

            if compareVersions(currentVersion, info.minSupported) == .orderedAscending {
                // Below minimum supported version
                latestVersion = info.latest
                downloadURL = URL(string: info.download.macos)
                unsupported = true
                updateAvailable = true
            } else if compareVersions(currentVersion, info.latest) == .orderedAscending {
                // Update available but not critical
                latestVersion = info.latest
                downloadURL = URL(string: info.download.macos)
                updateAvailable = true
            }
        } catch {
            // Fail silently — update check is best-effort
        }
    }

    /// Semantic version comparison
    private func compareVersions(_ a: String, _ b: String) -> ComparisonResult {
        let aParts = a.split(separator: ".").map { Int($0) ?? 0 }
        let bParts = b.split(separator: ".").map { Int($0) ?? 0 }
        let maxLen = max(aParts.count, bParts.count)

        for i in 0..<maxLen {
            let aVal = i < aParts.count ? aParts[i] : 0
            let bVal = i < bParts.count ? bParts[i] : 0
            if aVal < bVal { return .orderedAscending }
            if aVal > bVal { return .orderedDescending }
        }
        return .orderedSame
    }
}

// MARK: - JSON Model

private struct VersionInfo: Decodable {
    let latest: String
    let minSupported: String
    let download: Download

    struct Download: Decodable {
        let macos: String
        let windows: String
    }

    enum CodingKeys: String, CodingKey {
        case latest
        case minSupported = "min_supported"
        case download
    }
}
